### PR TITLE
Update ncm to ncm2

### PR DESCRIPTION
**Pull request by FatBoyXPC**
Note: this will break for anybody that currently uses ncm. I wasn't sure the right way to handle this - add a whole new group of stuff for ncm2, or just update it. I figured I would update it first since that was the easier path.

I tried to run the tests but `bundle install` failed, stating: `Could not locate Gemfile`

In order to get the routes completion working for ncm2, I had to make
the `capture()` function available from anywhere, and I needed to call it
with 2 arguments intead of 1.

### CODE REVIEW

File: plugin/laravel.vim [hunk 1]
```diff
@@ -101,23 +101,25 @@ augroup laravel_completion
   autocmd!
   " Set up nvim-completion-manager sources
   " :help NCM-source-examples
-  autocmd User CmSetup call cm#register_source({'name' : 'Laravel Route',
-        \ 'abbreviation': 'Route',
+  au User Ncm2Plugin call ncm2#register_source({
```
Comment (noahfrederick) [inline plugin/laravel.vim]: Here and elsewhere, please always spell out the full command name.

```suggestion
  autocmd User Ncm2Plugin call ncm2#register_source({
```

File: plugin/laravel.vim [hunk 2]
```diff
@@ -101,23 +101,25 @@ augroup laravel_completion
...
+        \ 'name' : 'Laravel Route',
```
Comment (noahfrederick) [inline plugin/laravel.vim]: Horizontal space in the popup menu is at a premium. While it's nice to make the source of the completion explicit, I settled on omitting "Laravel". Thoughts?

```suggestion
        \ 'name' : 'Route',
```

**Review by noahfrederick — CHANGES_REQUESTED** [review]
Thanks for taking this on!

> I wasn't sure the right way to handle this - add a whole new group of stuff for ncm2, or just update it.

Let's preserve the original ncm support, since the cost is low. But you can stick new autocommands in the same *group* (`laravel_completion`).

**Comment by noahfrederick** [comment]
> I tried to run the tests but `bundle install` failed, stating: `Could not locate Gemfile`

Not sure where this instruction came from, but there are none, unfortunately. :/

File: plugin/laravel.vim (hunk 1 above)
Comment (FatBoyXPC) [reply plugin/laravel.vim]: Understood! I haven't ever made a completion source and I was taking the shortest path to victory here, which meant copy/paste from the example in the docs and changing the necessary keys and values.hhhhh

File: plugin/laravel.vim (hunk 2 above)
Comment (FatBoyXPC) [reply plugin/laravel.vim]: The `name` isn't in the pop up menu, the `mark` is. It appears like `mark` is the new key for `abbreviation` - which honestly confuses me. I'm guessing `mark` has something to do with internal syntax to vim's pop up menu?

Here are some screenshots:

image

image

**Comment by FatBoyXPC** [comment]
That instruction came from `CONTRIBUTING.md`!

File: plugin/laravel.vim (hunk 2 above)
Comment (noahfrederick) [reply plugin/laravel.vim]: Got it—my mistake. Looks good.

**Comment by FatBoyXPC** [comment]
Made the requesting changes, let me know if there's anything else!

File: plugin/laravel.vim [hunk 3]
```diff
@@ -119,6 +119,29 @@ augroup laravel_completion
...
+        \ 'priority': 9,
+        \ 'subscope_enable': 1,
+        \ 'scope': ['php','blade'],
+        \ 'mark': 'Route',
+        \ 'word_pattern': '[A-Za-z0-9_.:-]+',
+        \ 'complete_pattern': ['\broute\([''"]'],
+        \ 'on_complete': 'laravel#completion#ncm2_routes',
+        \ })
+  autocmd User Ncm2Plugin call ncm2#register_source({
+        \ 'name' : 'Laravel View',
```
Comment (noahfrederick) [inline plugin/laravel.vim]:
```suggestion
        \ 'name': 'Laravel View',
```

File: plugin/laravel.vim [hunk 4]
```diff
@@ -119,6 +119,29 @@ augroup laravel_completion
         \ 'cm_refresh_patterns': ['\bview\([''"]', '@(component|extends|include)\([''"]'],
         \ 'cm_refresh': 'laravel#completion#cm_views',
         \ })
+
+  " Set up ncm2 sources
+  " :help ncm2#register_source-example
+  autocmd User Ncm2Plugin call ncm2#register_source({
+        \ 'name' : 'Laravel Route',
```
Comment (noahfrederick) [inline plugin/laravel.vim]:
```suggestion
        \ 'name': 'Laravel Route',
```

File: plugin/laravel.vim [hunk 5]
```diff
@@ -119,6 +119,29 @@ augroup laravel_completion
...
+        \ 'priority': 9,
+        \ 'subscope_enable': 1,
+        \ 'scope': ['php','blade'],
```
Comment (noahfrederick) [inline plugin/laravel.vim]:
```suggestion
        \ 'scope': ['php', 'blade'],
```

File: plugin/laravel.vim [hunk 6]
```diff
@@ -119,6 +119,29 @@ augroup laravel_completion
...
+        \ 'mark': 'Route',
+        \ 'word_pattern': '[A-Za-z0-9_.:-]+',
+        \ 'complete_pattern': ['\broute\([''"]'],
+        \ 'on_complete': 'laravel#completion#ncm2_routes',
+        \ })
+  autocmd User Ncm2Plugin call ncm2#register_source({
+        \ 'name' : 'Laravel View',
+        \ 'priority': 9,
+        \ 'subscope_enable': 1,
+        \ 'scope': ['php','blade'],
```
Comment (noahfrederick) [inline plugin/laravel.vim]:
```suggestion
        \ 'scope': ['php', 'blade'],
```

**Review by noahfrederick — CHANGES_REQUESTED** [review]
Looks good after these few formatting nitpicks.

I actually can't get it to work, but my autocompletion setup is perpetually broken in one way or another. As long as you've manually confirmed this works, I'm happy to merge it.

**Comment by FatBoyXPC** [comment]
It does work for me locally, but I wouldn't mind somebody else trying. Do you know anybody else using ncm2? How is your autocomplete currently broken? Maybe we could fix it and test my changes?

**Comment by noahfrederick** [comment]
Okay, I was able to create a working minimal configuration. Strangely, ncm2 is including view names (as well as the route names) in the route context and vice versa.

Screen Shot 2019-07-20 at 9 39 28 PM

**Comment by FatBoyXPC** [comment]
That's strange, that doesn't happen for me. Can you share your minimal configuration so I can also try it?

**Comment by noahfrederick** [comment]
Replace paths to local copies of plug-ins as needed:

```vim
call plug#begin()
call plug#('roxma/nvim-yarp')
call plug#('ncm2/ncm2')

call plug#('~/src/vim-composer')
call plug#('~/src/vim-laravel')
call plug#end()

augroup init_ncm2
  autocmd!
  autocmd BufEnter * call ncm2#enable_for_buffer()
augroup END

set completeopt=menuone         " Display menu when completing, even for only one candidate
set completeopt+=noselect       " Do not automatically select first item
set completeopt+=noinsert       " Do not automatically insert first item
```

Starting Neovim 0.3.8 in a Laravel project with:

```sh
nvim -u path/to/minimal.vim app/Http/Controllers/Controller.php
```

**Comment by noahfrederick** [comment]
I've been reading through the documentation. I think the `complete_pattern` key doesn't scope the completions in the way the original ncm did, but rather just provides extra contexts for automatically popping up the menu. I think we may need to check the context ourselves before returning results for a particular completion type, which is a little bit of a drag.

**Comment by FatBoyXPC** [comment]
Hm, I do see this bug now, but only after I call `route()` a 2nd time. I'm attempting the minimal configuration right now, but I'm getting a lot of errors. Did you do a `:PlugClean` and `:PlugInstall` with that config before attempting this?

**Comment by noahfrederick** [comment]
> Did you do a :PlugClean and :PlugInstall with that config before attempting this?

I did not, since it's a strict subset of my normal config. It shouldn't make a difference in my case.

> but I'm getting a lot of errors.

Can you tell from what script?

**Comment by FatBoyXPC** [comment]
The only difference is that I commented out the vim-composer line because I don't use it (and it's not necessary for what I use this plugin for). 

Here's a gif of what's happening:

2019-07-20_22:33:00_fatboyxpc

**Comment by FatBoyXPC** [comment]
Hey, I found this in the docs:

```
    complete_length                 ncm2-complete_length                        
                        The minimum length of the matching word for auto        
                        triggering popup menu.                                  
                        If it contains a negative value, completion will only   
                        be triggered by complete_pattern.
```

This part being important:

> If it contains a negative value, completion will only be triggered by complete_pattern.

Care to pull my changes and see if this is fixed?

**Review by noahfrederick — APPROVED** [review]
Works as expected now!

**Pull request merged**